File: 03_reusable.rb
require 'net/https'
require 'uri'
require 'json'

get_from_api = ->(base_url) {
  ->(endpoint) {
    ->(callback) {
      begin
        uri_str = "#{base_url}#{endpoint}"
        uri = URI.parse(uri_str)

        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE

        request = Net::HTTP::Get.new(uri.request_uri)
        response = http.request(request)

        callback.call(JSON.parse(response.body))
      rescue StandardError => e
        puts e.message
      end
    }
  }
}

get_github = get_from_api.('https://api.github.com')

get_github_users = get_github.('/users')
get_github_repos = get_github.('/repositories')

puts get_github_users.(->(data) { data.map { |u| u['login'] } })
puts get_github_users.(->(data) { data.map { |u| u['avatar_url'] } })

File: 03_reusable.js
const fetch = require('node-fetch')

const getFromApi = baseUrl => endpoint => callback =>
  fetch(`${baseUrl}${endpoint}`)
    .then(res => res.json())
    .then(data => callback(data))
    .catch(err => {
      console.error(err.message)
    })

const getGithub = getFromApi('https://api.github.com')

const getGithubUsers = getGithub('/users')
const getGithubRepos = getGithub('/repositories')

const prop = key => obj => obj[key]

const propLogin     = prop('login')
const propAvatarUrl = prop('avatar_url')

getGithubUsers(data => {
  console.log(data.map(propLogin))
})

getGithubUsers(data => console.log(data.map(propAvatarUrl)))
